src/G2/Equiv/Tactics.hs

```diff
@@ -942,26 +1062,34 @@ replaceMoreRestrictiveSubExpr' solver ns lemma@(Lemma { lemma_lhs = lhs_s, lemma
         altBind (Alt (DataAlt _ is) _) = map idName is
         altBind _ = []
 
+-- This is a looser version of the lemma soundness check from the paper.
+-- Instead of checking that the expression receiving the substitution is
+-- a suitable function application at the outermost level, we simply look
+-- for a sub-expression that fits that mold.  Substitutions can only
+-- happen within sub-expressions that satisfy the conditions.
```

Comment:
Can we add an argument/justification that loosening the requirement is an ok thing to do?